crates/red_knot/src/db.rs

```diff
@@ -34,6 +34,8 @@ pub trait SemanticDb: SourceDb {
     fn add_module(&mut self, path: &Path) -> Option<(Module, Vec<Arc<ModuleData>>)>;
 
     fn set_module_search_paths(&mut self, paths: Vec<ModuleSearchPath>);
+
+    fn eval_symbol(&mut self, file_id: FileId, symbol_id: SymbolId) -> Type;
```

Comment:
I think this should be a query and not a mutation
* Move up to other queries
* Change `self` to take a readonly reference.

Reply:
yes agreed, I'll make this update, and use internal mutability on the type store for the caching instead

Reply:
Going to do this in a follow up